storage_service, repair: use per-shard abort_source

bhalevy opened the pull request:
Prevent copying shared_ptr across shards
in do_sync_data_using_repair by allocating
a shared_ptr<abort_source> per shard in
node_ops_meta_data and respectively in node_ops_info.

Fixes #11826

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy commented:
Cc @Deexie

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2764/

bhalevy commented:
Sorry, I had a fix for the link failure that I forgot to amend to the patch I pushed.
Fixed in [6c3d90b](https://github.com/scylladb/scylladb/commit/6c3d90b7797ac9fcc1444e7a7ca3397db192680e)

bhalevy commented:
@Deexie with the task_manager infrastructure I think we can simplify the design by having the per-shard abort_source on the task impl holding today's repair_info (rather than a shared_ptr<abort_source>) and then using the tasks tree structure to request_abort, rather than storage_service::node_ops_abort which iterates on the node_ops_meta_data abort_sources.
If node_ops_meta_data / node_ops_info become tasks than the could subscribe a function to their own abort_source to pass the abort event down to subordinate tasks.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2765/

denesb commented on repair/repair.hh:
```diff
@@ -69,7 +69,7 @@ std::ostream& operator<<(std::ostream& os, const repair_uniq_id& x);
+    std::vector<shared_ptr<abort_source>> sharded_abort_source;
     std::list<gms::inet_address> ignore_nodes;
     void check_abort();
 };
```
Either delete the copy/move constructors and assignment operators or change the type to `std::vector<foreign_ptr<sharerd_ptr<abort_source>>>`. Maybe we should do both to be on the safe side.

bhalevy replied:
yikes :)
I'm tempted to simplify the abort design as per https://github.com/scylladb/scylladb/pull/11827#issuecomment-1288404065

bhalevy replied:
I wanted to change as little as possible in anticipation for https://github.com/scylladb/scylladb/pull/11597
and then adding support for abort on top of it.

denesb replied:
I think deleting the move/copy constructors and assignment operators is a small change and it goes a long way in avoiding accidents.

bhalevy commented:
In v2 ([0336da5](https://github.com/scylladb/scylladb/commit/0336da531cd6f47998769e928d8335a319c96f99)):
- delete the node_ops_info copy and move constructors
- use sharded<abort_source> for the per-shard repair_info
  - use abort_source* instead of shared_ptr<abort_source> in repair_info

denesb commented on repair/repair.hh:
```diff
@@ -71,6 +71,11 @@ struct node_ops_info {
+
+    node_ops_info(utils::UUID ops_uuid_, shared_ptr<abort_source> as_, std::list<gms::inet_address>&& ignore_nodes_) noexcept;
+    node_ops_info(const node_ops_info&) = delete;
+    node_ops_info(node_ops_info&&) = delete;
```
Does this also delete assignment ops? I don't recall the rules.

bhalevy replied:
yes, it does.

denesb commented on repair/repair.hh:
```diff
@@ -67,16 +67,28 @@ struct repair_uniq_id {
 
+private:
+    optimized_optional<abort_source::subscription> _abort_subscription;
+    sharded<abort_source> _sas;
```
I dont understand why this is sharded if the wrapping class is alreasy too.

bhalevy replied:
this is sharded while the wrapping class is shared.
god is in the details :)

denesb replied:
The start() and stop() methods confused me into thinking `node_ops_info` is `sharded<>` too.
I think these are weird sounding methods to have on an `info` class, but its fine. You mentioned this code is subject to further changes, maybe we can revisit this code after that and see if it needs some shaking-up.

bhalevy replied:
Yes.  I actually tried a different approach that would be closer to where we want to get in https://github.com/bhalevy/scylla/commit/eed170f35e7ea6b3d1d095bf448bb582d7f8845a
but I decided that the churn is too high for now.

denesb replied:
Ok, I'll queue once CI approves.

denesb reviewed (commented):
(sorry I pressed approve too soon)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2774/

bhalevy commented:
In v3 ([2954b13](https://github.com/scylladb/scylladb/commit/2954b138811723616c6f4aebbe75605c72712c2a)):
- refactored patchset for smaller independent chages
  - storage_service: node_ops_meta_data: add start and stop methods
  - repair: node_ops_info: add start and stop methods
- storage_service: node_ops_abort_thread: co_return only after printing log message
  - new, fixes a minor bug: missing printout on stop
- storage_service: node_ops_abort: pass null uuid to abort all node ops
  - fixing the failure seen in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2774/
  - by aborting and stopping all node ops metas in storage_service::stop

Deexie commented on repair/repair.cc:
```diff
@@ -568,6 +582,7 @@ repair_info::repair_info(repair_service& repair,
+    , as(asp)
     , _hints_batchlog_flushed(std::move(hints_batchlog_flushed)) {
     if (as != nullptr) {
         _abort_subscription = as->subscribe([this] () noexcept { abort(); });
```
Do we still need `_abort_subscription`and `aborted` flag in `repair_info`?
We could just check `as->abort_requested()`, since abort_source pointer is kept in repair_info.

bhalevy replied:
This is how the abort trigger is pased to repair_info. Otherwise, the ops info layer could call the abort method directly, if it would keep a foreign shared prr to the repair info (like children tasks) but that's a bigger change

bhalevy replied:
the repair info abort source is optional. I'm not sure if we can grt rid of the abort member

bhalevy replied:
@Deexie just to make it clear that after converting to the task manager infrastructure, all tasks should be abortable internally, those that are nonabortable are so just from the api perspective.
We can keep an abort_source at the generic task impl and request abort via the task class.
Then, the task impl replacing repair info can use it as you suggested above.

With that we'll be able to abort individual tasks, like a particular node op, by traversing the tree and aborting the descendents individually.

When a global abort is requested, e.g. on shutdown, we subscribe to at the module level, which should abort all tasks it owns by using the task map.  Similar to the meta layer in storage service in this series.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2780/

xemul commented on repair/repair.cc:
```diff
@@ -49,6 +49,12 @@
 logging::logger rlogger("repair");
 
+node_ops_info::node_ops_info(utils::UUID ops_uuid_, shared_ptr<abort_source> as_, std::list<gms::inet_address>&& ignore_nodes_) noexcept
+    : ops_uuid(ops_uuid_)
```
UUID copying is not noexcept

bhalevy replied:
Why not?
The following passes for me:
```c++
diff --git a/utils/uuid.cc b/utils/uuid.cc
index 0310a7add..1fbc2cb67 100644
--- a/utils/uuid.cc
+++ b/utils/uuid.cc
@@ -19,6 +19,15 @@
 
 namespace utils {
 
+static_assert(std::is_nothrow_default_constructible_v<UUID>);
+static_assert(std::is_nothrow_copy_constructible_v<UUID>);
+static_assert(std::is_nothrow_move_constructible_v<UUID>);
+
+using bogus_id = tagged_uuid<struct bogus_id_tag>;
+static_assert(std::is_nothrow_default_constructible_v<bogus_id>);
+static_assert(std::is_nothrow_copy_constructible_v<bogus_id>);
+static_assert(std::is_nothrow_move_constructible_v<bogus_id>);
+
 UUID
 make_random_uuid() noexcept {
     static thread_local std::mt19937_64 engine(std::random_device().operator()());
```

xemul replied:
Ouch, indeed. It's a pair of uint64_t's

xemul commented on service/storage_service.cc:
```diff
@@ -3755,15 +3773,15 @@ future<> storage_service::node_ops_abort_thread() {
-            }
             try {
-                co_await node_ops_abort(*uuid_opt);
+                co_await node_ops_abort(uuid_opt.value_or(utils::null_uuid()));
```
This seems not to be the part of just "fix cross-shard abort source usage". Would you extend the motivation in the commit message, please?

bhalevy replied:
sure

xemul commented on repair/repair.hh:
```diff
@@ -175,6 +184,7 @@ public:
     int ranges_index = 0;
     repair_stats _stats;
     std::unordered_set<sstring> dropped_tables;
+    abort_source* as;
```
This looks pretty risky. The lifetime of node_ops_info (which owns the abort_source this points to) is not synchronized with the lifetime of repair_info. Or is it?

bhalevy replied:
It's used only in the constructor to subscribe the callback.
If repair_info is destroyed along with the subscription then it will unsubscribe itself.
There should also be no problem to stop the `sharded<abort_source>` at `node_ops_info` while repair_info is alive since the subscription use boost-intrusive auto_unlink and destroying the list should safely detach the members.

bhalevy replied:
We can also go with the approach in https://github.com/bhalevy/scylla/commit/eed170f35e7ea6b3d1d095bf448bb582d7f8845a
where the `repair_info` can be accessed via `node_ops_info` via a foreign shared ptr, so the latter can abort the former using its `abort()` method, similar to the way we plan to do it in the future using the task manager infrastructure.

bhalevy replied:
> > It's used only in the constructor to subscribe the callback.
> 
> Then why does it exist as a member of repair_info?
> 
> > If repair_info is destroyed along with the subscription then it will unsubscribe itself.
> > There should also be no problem to stop the sharded<abort_source> at node_ops_info while repair_info is alive since the subscription use boost-intrusive auto_unlink and destroying the list should safely detach the members.
> 
> Absolutely. I've no objection about subscription, it's autounsubscribeable. Just the `abort_source* as` pointer itself

Good point. The member can indeed be removed.

bhalevy commented:
In v4 ([0ea8250](https://github.com/scylladb/scylladb/commit/0ea8250e83acc6e13e8b93afb6c911341349a307)):
- renamed and improved commit message of "storage_service: node_ops_abort_thread: abort all node ops on shutdown"
- got rid of `repair_info::abort_source*`
- rebased + retested

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2785/

bhalevy commented:
@scylladb/scylla-maint please merge

denesb commented on service/storage_service.cc:
```diff
@@ -3728,6 +3728,24 @@ future<> storage_service::node_ops_done(utils::UUID ops_uuid) {
+        for (auto it = _node_ops.begin(); it != _node_ops.end(); it = _node_ops.erase(it)) {
+            node_ops_meta_data& meta = it->second;
+            co_await meta.stop();
+        }
```
This is duplicating the stop code below.  We should extract it to a method or lambda to avoid accidental divergence.

bhalevy replied:
This path does 2 passes: first it aborts all repairs, then it awaits their stop() so they are slightly different.
In any case, this area will is being converted to use the task manager infrastructure in #11597 so I don't think it's worth the complication.

denesb replied:
The first half of the code is shared. But if there are further changes coming, then fine.

denesb reviewed (approved):
One minor comment, LGTM otherwise.
